airflow/jobs/local_task_job.py

```diff
@@ -95,6 +95,10 @@ def signal_handler(signum, frame):
                 # Monitor the task to see if it's done
                 return_code = self.task_runner.return_code()
                 if return_code is not None:
+                    if return_code != 0:
+                        msg = ("LocalTaskJob process exited with non zero "
+                               "status {}".format(return_code))
+                        raise AirflowException(msg)
```

Comment:
Throwing an exception here is heavier weight than we need. Just call on_fail here

Reply:
This PR is actually from previously a closed PR, in previous PR one of the reviewer recommend me to add an on_failure callback in base job. 
Here is the comment
https://github.com/apache/airflow/pull/1122#issuecomment-193252883.

I'm fine with either way, please let me know you thoughts

Reply:
No I did not resolve it... I can see the conversation is still here. I think it's just github mark it as outdated

Reply:
Back to your question,  we need to set the job state as FAILED as well. This is handled in BaseJob.run by throwing an exception.

Reply:
^^ @ashb